mentat/config.py

```diff
@@ -0,0 +1,174 @@
+        default=[
+            ["", "#9835bd"],
+            ["prompt", "#ffffff bold"],
+            ["continuation", "#ffffff bold"],
+        ],
+        metadata={
+            "description": "Styling information for the terminal.",
+            "only_config_file": True,
+        },
+    )
+
+    @classmethod
+    def add_fields_to_argparse(cls, parser: ArgumentParser) -> None:
+        for field in attr.fields(Config):
+            if "only_config_file" in field.metadata:
+                continue
+            name = [f"--{field.name.replace('_', '-')}"]
+            if "abbreviation" in field.metadata:
+                name.append(f"-{field.metadata['abbreviation'].replace('_', '-')}")
+
+            arguments = {
+                "help": field.metadata.get("description", ""),
+            }
+
+            if field.type == "bool":
+                if arguments.get("default", False):
+                    arguments["action"] = "store_false"
+                else:
+                    arguments["action"] = "store_true"
+            elif field.type == "int":
+                arguments["type"] = int
+            elif field.type == "float":
+                arguments["type"] = float
+            elif field.type == "list[str]":
+                arguments["nargs"] = "*"
+
+            parser.add_argument(*name, **arguments)
+
+    @classmethod
+    def create(cls, args: Namespace | None = None) -> Config:
+        config = Config()
+
+        # Each method overwrites the previous so they are in order of precedence
+        config.load_file(user_config_path)
+        config.load_file(get_shared_git_root_for_paths([Path(".")]) / config_file_name)
+
+        if args is not None:
+            config.load_namespace(args)
+
+        return config
+
+    def load_namespace(self, args: Namespace) -> None:
+        for field in vars(args):
+            if hasattr(self, field) and getattr(args, field) is not None:
+                setattr(self, field, getattr(args, field))
+
+    def load_file(self, path: Path) -> None:
+        if path.exists():
+            with open(path) as config_file:
+                try:
+                    config = json.load(config_file)
+                except JSONDecodeError:
+                    self.error(
+                        f"Warning: Config {path} contains invalid json; ignoring user"
+                        " configuration file"
+                    )
+                    return
+            for field in config:
+                if hasattr(self, field):
+                    setattr(self, field, config[field])
+                else:
+                    self.error(
+                        f"Warning: Config {path} contains unrecognized setting: {field}"
+                    )
+
+    def error(self, message: str) -> None:
+        self._errors.append(message)
+        try:
+            self.send_errors_to_stream()
+        except LookupError:
+            pass
+
+    def send_errors_to_stream(self):
```

Comment:
I agree this error stuff feels pretty weird; I noticed you call this function when the Session is created (right after SessionContext is set); why don't we just only call this function there? That way we don't need any of this except LookupError stuff.

Reply:
I did it this way because the config might have errors later in it's life cycle when the stream is initialized. This current version can't but when there's a `/config` command there could be type or setting dne errors.

Reply:
I don't really like the try/catch though. Mind if I set `SESSION_CONTEXT` to none right after it's declaration and then check that instead of catching an exception?